tests/test_butterworth_filter.py

```diff
@@ -0,0 +1,192 @@
+"""Tests for the Butterworth time domain filter"""
+import numpy as np
+import pytest
+from scipy import signal
+from eitprocessing.filters.butterworth_filters import BandPassFilter
+from eitprocessing.filters.butterworth_filters import BandStopFilter
+from eitprocessing.filters.butterworth_filters import ButterworthFilter
+from eitprocessing.filters.butterworth_filters import HighPassFilter
+from eitprocessing.filters.butterworth_filters import LowPassFilter
+
+
+@pytest.fixture
+def filter_arguments():
+    return {
+        "filter_type": "lowpass",
+        "cutoff_frequency": 10,
+        "sample_frequency": 50.2,
+        "order": 5,
+    }
+
+
+def check_filter_attributes(filter_, kwargs):
+    for key, value in kwargs.items():
+        assert getattr(filter_, key) == value
+
+    assert hasattr(filter_, "apply_filter")
+    assert callable(filter_.apply_filter)
+    assert filter_.available_in_gui
+
+
+def test_create_butterworth_filter(filter_arguments):
+    filter_ = ButterworthFilter(**filter_arguments)
+    check_filter_attributes(filter_, filter_arguments)
+
+
+def test_butterworth_range(filter_arguments):
+    # Tests whether an out-of-range order raises AttributeError
+    del filter_arguments["order"]
+
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments, order=11)
+
+    # Tests whether an explicit out-of-range order does not raise AttributeError
+    try:
+        filter_ = ButterworthFilter(**filter_arguments, order=11, override_order=True)
+    except AttributeError:
+        pytest.fail("Unexpected AttributeError")
+    assert filter_.order == 11
+
+
+def test_butterworth_filter_type(filter_arguments):
+    filter_arguments["filter_type"] = "invalid"
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments)
+
+
+def test_butterworth_cutoff_frequency(filter_arguments):
+    del filter_arguments["cutoff_frequency"]
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency="not a number")
+
+    filter_arguments["filter_type"] = "bandpass"
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency="not a number")
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=10)
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=("not a number", True))
+
+    with pytest.raises(TypeError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=[20, 30])
+
+    with pytest.raises(ValueError):
+        ButterworthFilter(**filter_arguments, cutoff_frequency=(1,))
+
+    with pytest.raises(ValueError):
```

Comment:
After writing below comments, I figured maybe it would be even nicer to create a helper function that is called in each specific attribute test. Something like:

```python
def check_inputs(
    filter_: ButterworthFilter, 
    filter_arguments: dict,
    test_attr: str, 
    invalid_inputs: list,
    valid_inputs: list,
    expected_error: Exception = TypeError
):
    for invalid in invalid_inputs:
        filter_arguments[test_attr] = invalid
        with pytest.raises(expected_error):
            filter_(**filter_arguments)
    
    for valid in valid_inputs:
        filter_arguments[test_attr] = valid
        try:
            filter_(**filter_arguments)
        except expected_error:
            pytest.fail("Unexpected error")
```

Note that it is untested, so might not be exactly right. Also, need to think of best way to test different error types (maybe best is just to call it multiple times, once for each error type)

Reply:
This is a good suggestion, maybe even for a broader application if generalised. Something like below (not tested).



```python

def check_inputs(
    class_: type, 
    init_arguments: dict,
    test_attr: str, 
    invalid_inputs: list,
    valid_inputs: list,
    expected_error: Exception,
    unexpected_errors: List[Exception]
):
    init_arguments = dict(init_arguments)  # create a copy of the init_arguments
    for invalid in invalid_inputs:
        init_arguments[test_attr] = invalid
        with pytest.raises(expected_error):
            class_(**filter_arguments)
    
    for valid in valid_inputs:
        init_arguments[test_attr] = valid
        try:
            filter_(**filter_arguments)
        except unexpected_errors:
            pytest.fail("Unexpected error")
```

It could then be used for any class. This function only tests one `test_attr` at a time, so would not really be suitable for different values of `cutoff_frequency` for different filter types, but that's easily solved by calling it twice: once for low/high pass, once for band pass/stop.

Do you agree we should create a general helper function for tests like this? I think that falls outside the scope of this PR.

Reply:
agreed, also that it's outside the scope